### ISSUE TITLE
`read_sample` gives `UnexpectedEof` error in 0.9

### ISSUE THREAD

**Issue by probablykasper** [alfg/mp4-rust#68]
Just tried updating to 0.9, but whenever I call `read_sample` I'm getting this error:
```
Err(IoError(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" }))
```

The mpeg_aac_decoder example seems to work fine, so there might be something I could do differently, but still this did work fine in 0.8.

Here's the branch where I tried the upgrade: https://github.com/probablykasper/redlux/tree/mp4-0.9
commit: https://github.com/probablykasper/redlux/commit/cc8add163a7fbf6009798ddab0722e9f6e8b94d7

**Comment by NightlySide** [comment]
It seems that the piece of code raising the error is the following one:

https://github.com/alfg/mp4-rust/blob/56cc260a5c103c66167f3c330be09da0b3e431ac/src/track.rs#L526-L535

**Comment by alfg** [comment]
Thanks for catching this. Will take a look as soon as I can.

**Comment by Serial-ATA** [comment]
@alfg I think I found the issue

https://github.com/alfg/mp4-rust/blob/93dbb564642cbbc577e706afdff87b8237cdbe4c/src/mp4box/tfhd.rs#L61

The base data offset is always assumed to be present, but I checked how FFmpeg handles this box:

https://github.com/FFmpeg/FFmpeg/blob/2b2380c098569af88e7a77aff78f330d11090512/libavformat/mov.c#L4819-L4821

The base data offset is dependent on the flags. I checked the asset in https://github.com/probablykasper/redlux/issues/4, which had an implicit offset, so the offset was actually being read from the stsd id and default duration, making it incredibly large (4294968320, greater than the file size).

I can make a PR to fix this, but I'm not sure how you would want this done. Since you're using the `ReadBox` impl I can't pass along the `moof` offset.

**Comment by alfg** [comment]
Hey @Serial-ATA, thanks for checking into this. Sorry, I haven't had much time lately to work on this, but I'm open to suggestions and PRs.

**Comment by Serial-ATA** [comment]
I'll have to either remove the `ReadBox` impl or somehow return the flags from it, maybe by storing them in the `TfhdBox` itself? What do you think?

**Comment by probablykasper** [comment]
@Serial-ATA any clue why it worked before 0.9?

**Comment by Serial-ATA** [comment]
Could be https://github.com/alfg/mp4-rust/pull/46? Not sure. A quick look through the history didn't turn up anything obvious, and I don't really know the project. :smile:

**Comment by w-flo** [comment]
I suspect this might be fixed in git master thanks to #86 and #106. I don't have a file that is affected – maybe someone in here could test this?

**Comment by probablykasper** [comment]
It seems to be fixed, yeah!
